File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/EncryptionUtils.java
package im.status.ethereum.module;

import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.Callback;
import android.util.Log;
import statusgo.Statusgo;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.function.Function;
import android.app.Activity;
import android.view.WindowManager;
import android.os.Build;
import android.view.Window;
import android.preference.PreferenceManager;
import android.content.SharedPreferences;

public class EncryptionUtils extends ReactContextBaseJavaModule {
    private static final String TAG = "EncryptionUtils";

    private ReactApplicationContext reactContext;
    private Utils utils;

    public EncryptionUtils(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.utils = new Utils(reactContext);
    }

    @Override
    public String getName() {
        return "EncryptionUtils";
    }

    @ReactMethod
    private void initKeystore(final String keyUID, final Callback callback) throws JSONException {
        Log.d(TAG, "initKeystore");

        final String commonKeydir = this.utils.pathCombine(this.utils.getNoBackupDirectory(), "/keystore");
        final String keydir = this.utils.pathCombine(commonKeydir, keyUID);

        StatusBackendClient.executeStatusGoRequestWithCallback(
            "InitKeystore",
            keydir,
            () -> Statusgo.initKeystore(keydir),
            callback
        );
    }

    @ReactMethod
    public void reEncryptDbAndKeystore(final String keyUID, final String password, final String newPassword, final Callback callback) throws JSONException {
        JSONObject params = new JSONObject();
        params.put("keyUID", keyUID);
        params.put("oldPassword", password);
        params.put("newPassword", newPassword);
        String jsonParams = params.toString();
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "ChangeDatabasePasswordV2",
            jsonParams,
            () -> Statusgo.changeDatabasePasswordV2(jsonParams),
            callback
        );
    }

    @ReactMethod
    public void convertToKeycardAccount(final String keyUID, final String accountData, final String options, final String keycardUID, final String password,
                                        final String newPassword, final Callback callback) throws JSONException {
        final String keyStoreDir = this.utils.getKeyStorePath(keyUID);
        JSONObject params = new JSONObject();
        params.put("keyUID", keyUID);
        params.put("account", new JSONObject(accountData));
        params.put("settings", new JSONObject(options));
        params.put("keycardUID", keycardUID);
        params.put("oldPassword", password);
        params.put("newPassword", newPassword);
        final String jsonParams = params.toString();
        StatusBackendClient.executeStatusGoRequest(
            "InitKeystore",
            keyStoreDir,
            () -> Statusgo.initKeystore(keyStoreDir)
        );
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "ConvertToKeycardAccountV2",
            jsonParams,
            () -> Statusgo.convertToKeycardAccountV2(jsonParams),
            callback
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String encodeTransfer(final String to, final String value) {
        try {
            JSONObject params = new JSONObject();
            params.put("to", to);
            params.put("value", value);
            String jsonParams = params.toString();
            return StatusBackendClient.executeStatusGoRequestWithResult(
                "EncodeTransferV2",
                jsonParams,
                () -> Statusgo.encodeTransferV2(jsonParams)
            );
        } catch (JSONException e) {
            Log.e(TAG, "Error creating JSON for encodeTransfer: " + e.getMessage());
            return null;
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String encodeFunctionCall(final String method, final String paramsJSON) {
        try {
            JSONObject params = new JSONObject();
            params.put("method", method);
            params.put("paramsJSON", new JSONObject(paramsJSON));
            String jsonString = params.toString();
            return StatusBackendClient.executeStatusGoRequestWithResult(
                "EncodeFunctionCallV2",
                jsonString,
                () -> Statusgo.encodeFunctionCallV2(jsonString)
            );
        } catch (JSONException e) {
            Log.e(TAG, "Error creating JSON for encodeFunctionCall: " + e.getMessage());
            return null;
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String decodeParameters(final String decodeParamJSON) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "DecodeParameters",
            decodeParamJSON,
            () -> Statusgo.decodeParameters(decodeParamJSON)
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String hexToNumber(final String hex) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "HexToNumber",
            hex,
            () -> Statusgo.hexToNumber(hex)
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String numberToHex(final String numString) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "NumberToHex",
            numString,
            () -> Statusgo.numberToHex(numString)
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String sha3(final String str) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "Sha3",
            str,
            () -> Statusgo.sha3(str)
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String utf8ToHex(final String str) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "Utf8ToHex",
            str,
            () -> Statusgo.utf8ToHex(str)
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String hexToUtf8(final String str) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "HexToUtf8",
            str,
            () -> Statusgo.hexToUtf8(str)
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String serializeLegacyKey(final String publicKey) {
        return StatusBackendClient.executeStatusGoRequestWithResult(
            "SerializeLegacyKey",
            publicKey,
            () -> Statusgo.serializeLegacyKey(publicKey)
        );
    }

    @ReactMethod
    public void setBlankPreviewFlag(final Boolean blankPreview) {
        final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this.reactContext);
        sharedPrefs.edit().putBoolean("BLANK_PREVIEW", blankPreview).commit();
        setSecureFlag();
    }

    private void setSecureFlag() {
        final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this.reactContext);
        final boolean setSecure = sharedPrefs.getBoolean("BLANK_PREVIEW", true);
        final Activity activity = this.reactContext.getCurrentActivity();
        if (activity != null) {
            activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    final Window window = activity.getWindow();
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH && setSecure) {
                        window.addFlags(WindowManager.LayoutParams.FLAG_SECURE);
                    } else {
                        window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE);
                    }
                }
            });
        }
    }

    @ReactMethod
    public void hashTransaction(final String txArgsJSON, final Callback callback) throws JSONException {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "HashTransaction",
            txArgsJSON,
            () -> Statusgo.hashTransaction(txArgsJSON),
            callback
        );
    }

    @ReactMethod
    public void hashMessage(final String message, final Callback callback) throws JSONException {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "HashMessage",
            message,
            () -> Statusgo.hashMessage(message),
            callback
        );
    }

    @ReactMethod
    public void multiformatDeserializePublicKey(final String multiCodecKey, final String base58btc, final Callback callback) throws JSONException {
        JSONObject params = new JSONObject();
        params.put("key", multiCodecKey);
        params.put("outBase", base58btc);
        String jsonParams = params.toString();
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "MultiformatDeserializePublicKeyV2",
            jsonParams,
            () -> Statusgo.multiformatDeserializePublicKeyV2(jsonParams),
            callback
        );
    }

    @ReactMethod
    public void deserializeAndCompressKey(final String desktopKey, final Callback callback) throws JSONException {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "DeserializeAndCompressKey",
            desktopKey,
            () -> Statusgo.deserializeAndCompressKey(desktopKey),
            callback
        );
    }

    @ReactMethod
    public void hashTypedData(final String data, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.hashTypedData(data), callback);
    }

    @ReactMethod
    public void hashTypedDataV4(final String data, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.hashTypedDataV4(data), callback);
    }

    @ReactMethod
    public void signMessage(final String rpcParams, final Callback callback) throws JSONException {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "SignMessage",
            rpcParams,
            () -> Statusgo.signMessage(rpcParams),
            callback
        );
    }

    @ReactMethod
    public void signTypedData(final String data, final String account, final String password, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.signTypedData(data, account, password), callback);
    }

    @ReactMethod
    public void signTypedDataV4(final String data, final String account, final String password, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.signTypedDataV4(data, account, password), callback);

    }

    @ReactMethod
    public void signGroupMembership(final String content, final Callback callback) throws JSONException {
        this.utils.executeRunnableStatusGoMethod(() -> Statusgo.signGroupMembership(content), callback);
    }


}

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/EncryptionUtils.kt
package im.status.ethereum.module

import android.app.Activity
import android.os.Build
import android.preference.PreferenceManager
import android.util.Log
import android.view.WindowManager
import com.facebook.react.bridge.*
import org.json.JSONException
import org.json.JSONObject
import statusgo.Statusgo
import java.lang.Exception
import java.lang.RuntimeException

class EncryptionUtils(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private val utils = Utils(reactContext)
    
    companion object {
        private const val TAG = "EncryptionUtils"
    }

    override fun getName(): String = "EncryptionUtils"

    @ReactMethod
    @Throws(Exception::class)
    private fun initKeystore(keyUID: String, callback: Callback) {
        Log.d(TAG, "initKeystore")
        val commonKeydir = utils.pathCombine(utils.getNoBackupDirectory(), "/keystore")
        val keydir = utils.pathCombine(commonKeydir, keyUID)

        StatusBackendClient.executeStatusGoRequestWithCallback(
            "InitKeystore",
            keydir,
            { Statusgo.initKeystore(keydir) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun reEncryptDbAndKeystore(keyUID: String, password: String, newPassword: String, callback: Callback) {
        val params = JSONObject().apply {
            put("keyUID", keyUID)
            put("oldPassword", password)
            put("newPassword", newPassword)
        }
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "ChangeDatabasePasswordV2",
            params.toString(),
            { Statusgo.changeDatabasePasswordV2(params.toString()) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun convertToKeycardAccount(
        keyUID: String, 
        accountData: String, 
        options: String, 
        keycardUID: String, 
        password: String,
        newPassword: String, 
        callback: Callback
    ) {
        val keyStoreDir = utils.getKeyStorePath(keyUID)
        val params = JSONObject().apply {
            put("keyUID", keyUID)
            put("account", JSONObject(accountData))
            put("settings", JSONObject(options))
            put("keycardUID", keycardUID)
            put("oldPassword", password)
            put("newPassword", newPassword)
        }
        val jsonParams = params.toString()
        
        StatusBackendClient.executeStatusGoRequest(
            "InitKeystore",
            keyStoreDir,
            { Statusgo.initKeystore(keyStoreDir) }
        )
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "ConvertToKeycardAccountV2",
            jsonParams,
            { Statusgo.convertToKeycardAccountV2(jsonParams) },
            callback
        )
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun encodeTransfer(to: String, value: String): String? {
        return try {
            val params = JSONObject().apply {
                put("to", to)
                put("value", value)
            }
            StatusBackendClient.executeStatusGoRequestWithResult(
                "EncodeTransferV2",
                params.toString(),
                { Statusgo.encodeTransferV2(params.toString()) }
            )
        } catch (e: JSONException) {
            Log.e(TAG, "Error creating JSON for encodeTransfer: ${e.message}")
            null
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun encodeFunctionCall(method: String, paramsJSON: String): String? {
        return try {
            val params = JSONObject().apply {
                put("method", method)
                put("paramsJSON", JSONObject(paramsJSON))
            }
            StatusBackendClient.executeStatusGoRequestWithResult(
                "EncodeFunctionCallV2",
                params.toString(),
                { Statusgo.encodeFunctionCallV2(params.toString()) }
            )
        } catch (e: JSONException) {
            Log.e(TAG, "Error creating JSON for encodeFunctionCall: ${e.message}")
            null
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun decodeParameters(decodeParamJSON: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "DecodeParameters",
            decodeParamJSON,
            { Statusgo.decodeParameters(decodeParamJSON) }
        )

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun hexToNumber(hex: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "HexToNumber",
            hex,
            { Statusgo.hexToNumber(hex) }
        )

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun numberToHex(numString: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "NumberToHex",
            numString,
            { Statusgo.numberToHex(numString) }
        )

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun sha3(str: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "Sha3",
            str,
            { Statusgo.sha3(str) }
        )

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun utf8ToHex(str: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "Utf8ToHex",
            str,
            { Statusgo.utf8ToHex(str) }
        )

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun hexToUtf8(str: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "HexToUtf8",
            str,
            { Statusgo.hexToUtf8(str) }
        )

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun serializeLegacyKey(publicKey: String): String = 
        StatusBackendClient.executeStatusGoRequestWithResult(
            "SerializeLegacyKey",
            publicKey,
            { Statusgo.serializeLegacyKey(publicKey) }
        )

    @ReactMethod
    fun setBlankPreviewFlag(blankPreview: Boolean) {
        PreferenceManager.getDefaultSharedPreferences(reactContext)
            .edit()
            .putBoolean("BLANK_PREVIEW", blankPreview)
            .commit()
        setSecureFlag()
    }

    private fun setSecureFlag() {
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(reactContext)
        val setSecure = sharedPrefs.getBoolean("BLANK_PREVIEW", true)
        
        reactContext.currentActivity?.run {
            runOnUiThread {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH && setSecure) {
                    window.addFlags(WindowManager.LayoutParams.FLAG_SECURE)
                } else {
                    window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
                }
            }
        }
    }

    @ReactMethod
    @Throws(Exception::class)
    fun hashTransaction(txArgsJSON: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "HashTransaction",
            txArgsJSON,
            { Statusgo.hashTransaction(txArgsJSON) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun hashMessage(message: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "HashMessage",
            message,
            { Statusgo.hashMessage(message) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun multiformatDeserializePublicKey(multiCodecKey: String, base58btc: String, callback: Callback) {
        val params = JSONObject().apply {
            put("key", multiCodecKey)
            put("outBase", base58btc)
        }
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "MultiformatDeserializePublicKeyV2",
            params.toString(),
            { Statusgo.multiformatDeserializePublicKeyV2(params.toString()) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun deserializeAndCompressKey(desktopKey: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "DeserializeAndCompressKey",
            desktopKey,
            { Statusgo.deserializeAndCompressKey(desktopKey) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun hashTypedData(data: String, callback: Callback) {
        utils.executeRunnableStatusGoMethod({ Statusgo.hashTypedData(data) }, callback)
    }

    @ReactMethod
    @Throws(Exception::class)
    fun hashTypedDataV4(data: String, callback: Callback) {
        utils.executeRunnableStatusGoMethod({ Statusgo.hashTypedDataV4(data) }, callback)
    }

    @ReactMethod
    @Throws(Exception::class)
    fun signMessage(rpcParams: String, callback: Callback) {
        StatusBackendClient.executeStatusGoRequestWithCallback(
            "SignMessage",
            rpcParams,
            { Statusgo.signMessage(rpcParams) },
            callback
        )
    }

    @ReactMethod
    @Throws(Exception::class)
    fun signTypedData(data: String, account: String, password: String, callback: Callback) {
        utils.executeRunnableStatusGoMethod({ Statusgo.signTypedData(data, account, password) }, callback)
    }

    @ReactMethod
    @Throws(Exception::class)
    fun signTypedDataV4(data: String, account: String, password: String, callback: Callback) {
        utils.executeRunnableStatusGoMethod({ Statusgo.signTypedDataV4(data, account, password) }, callback)
    }

    @ReactMethod
    @Throws(Exception::class)
    fun signGroupMembership(content: String, callback: Callback) {
        utils.executeRunnableStatusGoMethod({ Statusgo.signGroupMembership(content) }, callback)
    }
}
